### ISSUE TITLE
[Inductor] Redundant computation in the max_pool2d_with_indices_backward decomposition for non-overlapping windows

### ISSUE THREAD

**Issue by orrangetabby17** [pytorch/pytorch#195123]
## Summary

The `max_pool2d_with_indices_backward` decomposition which is introduced in https://github.com/pytorch/pytorch/pull/167318 applies the generic
`scatter_add` formulation to both overlapping and non-overlapping pooling
windows:

~~~text
grad_output
    -> flatten
    -> scatter_add into grad_input
~~~

For non-overlapping pooling, each input position belongs to at most one output window. The decomposition nevertheless retains the destination initialization, indirect scatter, and accumulation machinery required by overlapping windows.

This introduces redundant work and produces inefficient generated kernels for
common configurations such as:

~~~python
kernel_size = [2, 2]
stride = [2, 2]
padding = [0, 0]
dilation = [1, 1]
~~~

On GPU backends, the generated scatter kernel may contain sparse atomic writes.

## Reproducer

```python
import torch

from torch._decomp import get_decompositions
from torch.fx.experimental.proxy_tensor import make_fx


aten = torch.ops.aten

decomp = get_decompositions(
    [aten.max_pool2d_with_indices_backward]
)[aten.max_pool2d_with_indices_backward.default]

x = torch.randn(8, 64, 224, 224).contiguous(
    memory_format=torch.channels_last
)

result, indices = aten.max_pool2d_with_indices.default(
    x,
    [2, 2],
    [2, 2],
    [0, 0],
    [1, 1],
    False,
)

grad_output = torch.randn_like(result)

graph = make_fx(
    lambda grad, inp, ind: decomp(
        grad,
        inp,
        [2, 2],
        [2, 2],
        [0, 0],
        [1, 1],
        False,
        ind,
    )
)(grad_output, x, indices)

print(graph.code)

assert any(
    node.target == aten.scatter_add.default
    for node in graph.graph.nodes
)
```

Graph:
```
    def call(self, args):
        getitem_1, tangents_1 = args
        args.clear()
        with torch.cuda._DeviceGuard(0):
            torch.cuda.set_device(0)
            buf0 = empty_strided_cuda((512, 50176), (50176, 1), torch.float32)
            # Topologically Sorted Source Nodes: [clone, view, max_pool2d, clone_1, view_1, full_default, scatter_add], Original ATen: [aten.max_pool2d_with_indices_backward, aten.max_pool2d_with_indices]
            raw_stream0 = get_raw_stream(0)
            triton_poi_fused_max_pool2d_with_indices_max_pool2d_with_indices_backward_0.run(buf0, 25690112, stream=raw_stream0)
            assert_size_stride_grouped((getitem_1, tangents_1), ((8, 64, 112, 112), (8, 64, 112, 112)), ((802816, 1, 7168, 64), (802816, 1, 7168, 64)), 'input')
            tangents_1 = copy_if_misaligned(tangents_1)
            # Topologically Sorted Source Nodes: [clone, view, max_pool2d, clone_1, view_1, full_default, scatter_add], Original ATen: [aten.max_pool2d_with_indices_backward, aten.max_pool2d_with_indices]
            raw_stream0 = get_raw_stream(0)
            triton_poi_fused_max_pool2d_with_indices_max_pool2d_with_indices_backward_1.run(getitem_1, tangents_1, buf0, 6422528, stream=raw_stream0)
            del getitem_1
            del tangents_1
            buf2 = empty_strided_cuda((8, 64, 224, 224), (3211264, 1, 14336, 64), torch.float32)
            # Topologically Sorted Source Nodes: [view_2, clone_2], Original ATen: [aten.max_pool2d_with_indices_backward]
            raw_stream0 = get_raw_stream(0)
            triton_poi_fused_max_pool2d_with_indices_backward_2.run(buf0, buf2, 512, 50176, stream=raw_stream0)
            del buf0
        return (buf2, )
```
Kernel0:
```
@triton.jit
def triton_(out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 25690112
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = tl.full([XBLOCK], True, tl.int1)[:]
    x0 = xindex
    tmp0 = tl.full([1], 0.0, tl.float32)
    tl.store(out_ptr0 + (x0), tmp0, None)

```
Kernel 1:
```
@triton.jit
def triton_(in_ptr0, in_ptr1, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 6422528
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = tl.full([XBLOCK], True, tl.int1)[:]
    x0 = (xindex % 512)
    x1 = xindex // 512
    tmp0 = tl.load(in_ptr0 + (64*x1 + 802816*(x0 // 64) + ((x0 % 64))), None)
    tmp6 = tl.load(in_ptr1 + (64*x1 + 802816*(x0 // 64) + ((x0 % 64))), None)
    tmp1 = (tl.full([XBLOCK], 4, tl.int32)).to(tl.int32)
    tmp2 = tmp0 + tmp1
    tmp3 = tmp0 < 0
    tmp4 = tl.where(tmp3, tmp2, tmp0)
    tl.device_assert((0 <= tmp4) & (tmp4 < 4), "index out of bounds: 0 <= tmp4 < 4")
    tl.atomic_add(out_ptr0 + (tmp4 + 2*((x1 % 112)) + 222*(tmp4 // 2) + 448*(x1 // 112) + 50176*x0), tmp6, None, sem='relaxed')
```
kernel2
```
@triton.jit
def triton_(in_ptr0, out_ptr0, ynumel, xnumel, YBLOCK : tl.constexpr, XBLOCK : tl.constexpr):
    ynumel = 512
    xnumel = 50176
    yoffset = tl.program_id(1) * YBLOCK
    yindex = yoffset + tl.arange(0, YBLOCK)[:, None]
    ymask = yindex < ynumel
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[None, :]
    xmask = xindex < xnumel
    x2 = xindex
    y3 = yindex
    y0 = (yindex % 64)
    y1 = yindex // 64
    tmp0 = tl.load(in_ptr0 + (x2 + 50176*y3), xmask & ymask)
    tl.store(out_ptr0 + (y0 + 64*x2 + 3211264*y1), tmp0, xmask & ymask)
```
## Redundant work

For non-overlapping pooling:
```
stride_h >= kernel_h
stride_w >= kernel_w
padding == 0
dilation == 1
```
each input coordinate `(h, w)` has at most one possible output coordinate:
The backward can instead be expressed as a dense input-centric operation:

```
for each input position:
    find its only possible output window
    read the saved max index
    compare it with the current input index
    write grad_output if selected, otherwise write zero
```

### Versions

```
Collecting environment information...
PyTorch version: 2.14.0.dev20260719+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: version 4.4.0
Libc version: glibc-2.35

Python version: 3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 16:50:00) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.19.0-32-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 13.0.88
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 4080 SUPER
GPU 1: NVIDIA GeForce RTX 4070 Ti SUPER

Nvidia driver version: 580.173.02
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Core(TM) Ultra 9 285K
CPU family:                      6
Model:                           198
Thread(s) per core:              1
Core(s) per socket:              1
Socket(s):                       24
Stepping:                        2
CPU max MHz:                     5100.0000
CPU min MHz:                     800.0000
BogoMIPS:                        7372.80
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single intel_ppin ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves avx_vnni wbnoinvd dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid bus_lock_detect movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       768 KiB (20 instances)
L1i cache:                       1.3 MiB (20 instances)
L2 cache:                        40 MiB (12 instances)
L3 cache:                        36 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==2.5.1
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.24.0.43
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.30.7
[pip3] nvidia-nvjitlink==13.3.33
[pip3] nvidia-nvtx==13.0.85
[pip3] optree==0.19.1
[pip3] torch==2.14.0.dev20260719+cu130
[pip3] torchbench==0.1
[pip3] torchvision==0.29.0.dev20260720+cu130
[pip3] triton==3.8.0+git4774139f
[conda] numpy                     2.5.1                    pypi_0    pypi
[conda] nvidia-cublas             13.1.1.3                 pypi_0    pypi
[conda] nvidia-cuda-cupti         13.0.85                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc         13.0.88                  pypi_0    pypi
[conda] nvidia-cuda-runtime       13.0.96                  pypi_0    pypi
[conda] nvidia-cudnn-cu13         9.24.0.43                pypi_0    pypi
[conda] nvidia-cufft              12.0.0.61                pypi_0    pypi
[conda] nvidia-curand             10.4.0.35                pypi_0    pypi
[conda] nvidia-cusolver           12.0.4.66                pypi_0    pypi
[conda] nvidia-cusparse           12.6.3.3                 pypi_0    pypi
[conda] nvidia-cusparselt-cu13    0.8.1                    pypi_0    pypi
[conda] nvidia-nccl-cu13          2.30.7                   pypi_0    pypi
[conda] nvidia-nvjitlink          13.3.33                  pypi_0    pypi
[conda] nvidia-nvtx               13.0.85                  pypi_0    pypi
[conda] optree                    0.19.1                   pypi_0    pypi
[conda] torch                     2.14.0.dev20260719+cu130          pypi_0    pypi
[conda] torchbench                0.1                      pypi_0    pypi
[conda] torchvision               0.29.0.dev20260720+cu130          pypi_0    pypi
[conda] triton                    3.8.0+git4774139f          pypi_0    pypi
```

cc @mikaylagawarecki @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo
